### PR TITLE
feat(avoidance): support s-shape avoidance

### PR DESCRIPTION
**Pull request by satoshi-ota**
### <samp>🤖 Generated by Copilot at aede2b8</samp>

Refactored the `AvoidanceModule` class and related utils to improve the logic and performance of shift line generation and processing for path planning. Introduced new data structures and functions to group and process shift lines based on detected objects and reference path. Simplified and modularized the code for better readability and quality.

https://github.com/autowarefoundation/autoware.universe/assets/44889564/6e52f8d0-ac60-45f6-871c-3b6aec5183fc

## Tests performed

- [x] [PASS TIER IV INTERNAL SCENARIOS](https://evaluation.tier4.jp/evaluation/reports/3e1e24d7-9fa3-56b0-b786-2aa705d6df88?project_id=prd_jt)

## Effects on system behavior

Make it possible to avoid objects in complex scene.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 1]
```diff
@@ -1221,99 +1281,183 @@ AvoidLineArray AvoidanceModule::extractShiftLinesFromLine(ShiftLineData & shift_
...
+
+    if (no_conflict(return_line, avoid_line)) {
+      ret.push_back(outlines.at(i));
+      continue;
+    }
+
+    const auto merged_shift_line = merge(return_line, avoid_line, getOriginalShiftLineUniqueId());
+
+    if (!isComfortable(AvoidLineArray{merged_shift_line})) {
+      ret.push_back(outlines.at(i));
```
Comment (rej55) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1322]: @satoshi-ota 
Is it okay to push back outline if `isComfortable` returns `false`?

Comment (satoshi-ota) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1322]: @rej55 Thanks for you comment!
This line checks feasibility of the merged shift line, not for the original avoid outline.
If the merged shift line is not feasible within accel/jerk constraints, skip merge process and use original avoid outline.

Comment (rej55) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1322]: @satoshi-ota Okay, I see. Thanks!

**Review by rej55 — APPROVED** [review]
LGTM

**Pull request merged**